### PR TITLE
[CLD-8507] Upgrade Go version, controller-runtime dependency

### PR DESCRIPTION
**Pull request by nickmisasi**
#### Summary
Updates controller-runtime dependency to latest. This requires a Go version upgrade. The Go version upgrade required a few other package upgrades. 

The controller-runtime upgrade resulted in some changes to the way the fake client handles status updates - see https://github.com/kubernetes-sigs/controller-runtime/pull/2259 - adjusted tests accordingly so they'd pass. 

#### Ticket Link
https://mattermost.atlassian.net/browse/CLD-8507

#### Release Note

```release-note
None
```

### CODE REVIEW

File: scripts/k8s.io/code-generator/generate-groups.sh [hunk 1]
```diff
@@ -65,42 +65,52 @@ for GVs in ${GROUPS_WITH_VERSIONS}; do
...
-      --input-dirs "$(codegen::join , "${FQ_APIS[@]}")" \
       -O zz_generated.deepcopy \
+      "$(codegen::join , "${FQ_APIS[@]}")" \
       "$@"
 fi
 
 if [ "${GENS}" = "all" ] || grep -qw "client" <<<"${GENS}"; then
   echo "Generating clientset for ${GROUPS_WITH_VERSIONS} at ${OUTPUT_PKG}/${CLIENTSET_PKG_NAME:-clientset}"
+  TEMP_DIR=$(mktemp -d)
+  trap 'rm -rf "${TEMP_DIR}"' EXIT
```
Comment (nickmisasi) [inline scripts/k8s.io/code-generator/generate-groups.sh:84]: --output-dir became a required param with the upgraded version - this just creates a temporary dir that's later deleted to use for that arg.

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -21,7 +21,7 @@ jobs:
           fetch-depth: 0
 
       - name: ci/setup-go
-        uses: actions/setup-go@6edd4406fa81c3da01a34fa6f6343087c207a568 # v3.5.0
+        uses: actions/setup-go@41dfa10bad2bb2ae585af6ee5bb4d7d973ad74ed # v3.5.0
```
Comment (gabrieljackson) [inline .github/workflows/ci.yml:24]: Can we get this version number updated to avoid possible future confusion? Same thing for the one below.

**Review by gabrieljackson — APPROVED** [review]
Looks great! Just one small thing to update.

File: controllers/mattermost/mattermost/controller_test.go [hunk 3]
```diff
@@ -177,15 +178,20 @@ func TestReconcile(t *testing.T) {
...
 		require.NoError(t, err)
 
 		t.Run("succeed if 0 replicas expected", func(t *testing.T) {
 			orgReplicas := mm.Spec.Replicas
 			replicasZero := int32(0)
 			mm.Spec.Replicas = &replicasZero
 
-			err = c.Update(context.TODO(), mm)
+			var fetchedMM mmv1beta.Mattermost
+			err = c.Get(context.Background(), mmKey, &fetchedMM)
```
Comment (fmartingr) [inline controllers/mattermost/mattermost/controller_test.go:190]:
```suggestion
			err = c.Get(context.Background(), mmKey, &fetchedMM)
			require.NoError(t, err)
```

**Review by fmartingr — APPROVED** [review]
👍  At some point it would be useful to unify the Golang version from the `go.mod` file and just pass it along to the files that require it.

**Pull request merged**